### ISSUE TITLE
Custom experience for specific users

### ISSUE THREAD

**Issue by brashrebel** [realbigplugins/client-dash-pro#21]
Right now the widgets and menus can be modified for _roles_. But a frequently suggested feature (though not always asked for in this way) would be the ability to define a custom menu or dashboard for a _user_ or users. This would be a **super** compelling feature for the pro version.

**Comment by brashrebel** [comment]
Would be way better for everyone when answering questions like these:
https://wordpress.org/support/topic/submenu-option-gone/#post-9914504
https://wordpress.org/support/topic/client-dash-appears-in-clients-menu/#post-9914486

**Comment by joelworsham** [comment]
I like it. How would you envision the editing experience for this within the Admin Customizer?

**Comment by brashrebel** [comment]
Couple ideas that come to mind:

A. Create a new interface where the admin can view available users and select which ones to create a custom experience for. The act of selecting them there makes them appear in the "Customizing:" select field at the bottom of the Tool Area.
B. Change the select field to be a search and select. Clicking it could reveal which custom experiences already exist and allow typing to search for either roles OR usernames.
C. Where the "Customize:" select field is now, put a toggle which switches between Roles and Users. Toggling that will change which list is available in the select field.

**Comment by joelworsham** [comment]
B is interesting...... What if rather than being strictly role based, and even strictly role/user based, it was "experience" based. You could create any number of "experiences" that you can then customize for. You can then add various things and conditions to each experience. For instance, users and/or roles. THis leaves room for flexibility.... whoah... like crazy room.

- User has sold X products in marketplace
- User has an email with a certain domain
- User has a specific subscription
- User belongs to X multisites
- User has been inactive for X days
- etc etc etc etc etc

**Comment by brashrebel** [comment]
I really like this direction! Couple thoughts hitting me right now:

1. Sounds like we would have to be really careful about multiple conditions being met. Users of CD could get confused quickly if they start applying experiences and more than one are applicable for another user. Like you give user eric and role editor both custom experiences but what happens if eric is an editor? I'm sure we can overcome it but it could get hairy.
2. We're starting to throw out some really awesome next level ideas. That shouldn't stop but we probably should be pragmatic and agree on where to start. This issue could get huge but I'd rather we just make it the first of a series of issues which build towards this grand vision. If you agree, what's that starting point? Is it "user based experiences" or "condition based experiences" or just shifting to the concept of "experiences" or something else?

**Comment by joelworsham** [comment]
1. This is big actually. Because, when using roles, it is super straightforward to generate the menus and previews for what should and should not be visible... This actually would be very difficult if roles and users were mixed... hmm...
2. Hmmm tough. I think individual users having custom experiences is the biggest here.

**Comment by brashrebel** [comment]
We have had three requests for this in two days (according to what I've seen come through in CD support) so I'm thinking it is super valid. We'd quadruple Pro sales easily with this feature.

**Comment by joelworsham** [comment]
Yup, I'm definitely on-board, just gotta iron out exactly how we'll approach this.

My vote is to table "custom experiences" and, for now, have Pro add the ability to customize for either a role or a user, one at a time.

**Comment by d4mation** [comment]
This would be useful in cases where a User has multiple Roles, like in bbPress. This was causing bbPress Menu Items to not show properly in the Menu Customizer because the Menu Customizer did not show any added Menu Items from having a bbPress Role. 

[I ended up addressing that for any mapped Roles in bbPress](https://github.com/realbigplugins/client-dash/commit/d3dc01953441982b4007b2d387472669417d02e5) (Keymaster -> Admin, Subscriber -> Participant, etc.) but if there were a User who is a Subscriber-user but was manually granted the Keymaster bbPress Role, the Menu Customizer wouldn't work properly in their case as it would only show any Subscriber-level menu items, plus bbPress Participant items, plus any customizations.

If these "Experiences" were to account for a User's Capabilities and not only ones for a particular Role that would prevent any multi-role oddities altogether. In fact, having "Experiences" being able to apply to specific Role Combinations could be cool, but I imagine outside of situations like bbPress not used too often.
